pkg/transformers/namereference.go

```diff
@@ -72,23 +72,55 @@ func (o *nameReferenceTransformer) updateNameReference(
 	backRef gvk.Gvk, m resmap.ResMap) func(in interface{}) (interface{}, error) {
 	return func(in interface{}) (interface{}, error) {
 		s, ok := in.(string)
-		if !ok {
-			return nil, fmt.Errorf("%#v is expected to be %T", in, s)
+		if ok {
+			for id, res := range m {
+				if id.Gvk().IsSelected(&backRef) && id.Name() == s {
+					matchedIds := m.FindByGVKN(id)
+					// If there's more than one match, there's no way
+					// to know which one to pick, so emit error.
+					if len(matchedIds) > 1 {
+						return nil, fmt.Errorf(
+							"Multiple matches for name %s:\n  %v", id, matchedIds)
+					}
+					// Return transformed name of the object,
+					// complete with prefixes, hashes, etc.
+					return res.GetName(), nil
+				}
+			}
+			return in, nil
 		}
-		for id, res := range m {
-			if id.Gvk().IsSelected(&backRef) && id.Name() == s {
-				matchedIds := m.FindByGVKN(id)
-				// If there's more than one match, there's no way
-				// to know which one to pick, so emit error.
-				if len(matchedIds) > 1 {
-					return nil, fmt.Errorf(
-						"Multiple matches for name %s:\n  %v", id, matchedIds)
+		l, ok := in.([]interface{})
```

Comment:
Probably we can change it to something like the following for better readability.
```
switch in.(type) {
case string:
  // process string
case []interface{}:
  // process array
default:
  // return error
}
```

Reply:
updated